### ISSUE TITLE
Same hyperparams/seed but different output?

### ISSUE THREAD

**Issue by manueltonneau** [ThilinaRajapakse/simpletransformers#907]
**Describe the bug**
I am training BERT-based models (in my case, `vinai/bertweet-base`) on a binary classification problem. I trained twice the same type of model on the same data, with the same hyperparameters and seed. I then load both "best_models" in terms of evaluation loss and evaluate them but the output/performance metrics are different... Is there another source of variation (apart from seed and hyperparams) that could explain this and that I am missing?

**To Reproduce**

```python
classification_args = {'train_batch_size': 8, 'overwrite_output_dir': True, 'evaluate_during_training': True,
                                      'save_model_every_epoch': True, 'save_eval_checkpoints': True,
                                      'output_dir': path_to_store_model, 'best_model_dir': path_to_store_best_model,
                                      'evaluate_during_training_verbose': True,
                                      'num_train_epochs': args.num_train_epochs, "use_early_stopping": True,
                                      "early_stopping_delta": 0, "early_stopping_metric": "eval_loss",
                                      "early_stopping_metric_minimize": True, "tensorboard_dir": f"runs/{args.slurm_job_id}_{name_val_file.replace('val_', '')}/" ,
                                      "manual_seed": 0}

model = ClassificationModel('bertweet', 'vinai/bertweet-base', num_labels=2, use_cuda=True,
                                args=classification_args)
eval_metrics = {
        "precision": sklearn.metrics.precision_score,
        "recall": sklearn.metrics.recall_score,
        "f1": sklearn.metrics.f1_score
    }
# Train the model
model.train_model(train_df, eval_df=eval_df, output_dir=path_to_store_model, **eval_metrics)

# Load best model (in terms of evaluation loss)
train_args = read_json(filename=os.path.join(path_to_store_best_model, 'model_args.json'))
best_model = ClassificationModel('bertweet', path_to_store_best_model, args=train_args)

# EVALUATION ON EVALUATION SET
result, model_outputs, wrong_predictions = best_model.eval_model(eval_df)
scores = np.array([softmax(element)[1] for element in model_outputs])
y_pred = np.vectorize(convert_score_to_predictions)(scores)
# Compute AUC
fpr, tpr, thresholds = metrics.roc_curve(eval_df['labels'], scores)
auc_eval = metrics.auc(fpr, tpr)
# Centralize evaluation results in a dictionary
slurm_job_timestamp = args.slurm_job_timestamp
slurm_job_id = args.slurm_job_id
eval_results_eval_set_dict = {'slurm_job_id': slurm_job_id,
                              'slurm_job_timestamp': slurm_job_timestamp,
                              'slurm_job_Berlin_date_time': str(datetime.fromtimestamp(int(slurm_job_timestamp),
                                                                                       tz=pytz.timezone(
                                                                                           'Europe/Berlin'))),
                              'model_type': args.model_type,
                              'evaluation_data_path': args.eval_data_path,
                              'precision': metrics.precision_score(eval_df['labels'], y_pred),
                              'recall': metrics.recall_score(eval_df['labels'], y_pred),
                              'f1': metrics.f1_score(eval_df['labels'], y_pred),
                              'auc': auc_eval
                              }

```

**Expected behavior**
The values for keys `precision`, `recall`, `f1` and `auc` should be equal across finetunings with same hyperparams/seed.

**Output of pip freeze**
absl-py==0.11.0
altair==4.1.0
argon2-cffi==20.1.0
astor==0.8.1
async-generator==1.10
attrs==20.2.0
backcall==0.2.0
base58==2.0.1
bleach==3.2.1
blinker==1.4
boto3==1.15.11
botocore==1.18.11
cachetools==4.1.1
certifi==2020.6.20
cffi==1.14.3
chardet==3.0.4
click==7.1.2
colorama==0.4.4
configparser==5.0.0
cycler==0.10.0
decorator==4.4.2
defusedxml==0.6.0
docker-pycreds==0.4.0
ekphrasis==0.5.1
emoji==0.6.0
entrypoints==0.3
enum-compat==0.0.3
filelock==3.0.12
ftfy==5.8
future==0.18.2
gast==0.2.2
gitdb==4.0.5
GitPython==3.1.9
google-pasta==0.2.0
grpcio==1.33.2
h5py==2.10.0
idna==2.10
importlib-metadata==2.0.0
ipykernel==5.3.4
ipython==7.18.1
ipython-genutils==0.2.0
ipywidgets==7.5.1
jedi==0.17.2
Jinja2==2.11.2
jmespath==0.10.0
joblib==0.16.0
jsonschema==3.2.0
jupyter-client==6.1.7
jupyter-core==4.6.3
jupyterlab-pygments==0.1.2
Keras==2.4.3
Keras-Applications==1.0.8
Keras-Preprocessing==1.1.2
kiwisolver==1.2.0
Markdown==3.3.3
MarkupSafe==1.1.1
matplotlib==3.3.2
mistune==0.8.4
nbclient==0.5.0
nbconvert==6.0.7
nbformat==5.0.7
nest-asyncio==1.4.1
nltk==3.5
notebook==6.1.4
numpy==1.19.1
opt-einsum==3.3.0
packaging==20.4
pandas==1.1.2
pandocfilters==1.4.2
parso==0.7.1
pathtools==0.1.2
pexpect==4.8.0
pickleshare==0.7.5
Pillow==7.2.0
prometheus-client==0.8.0
promise==2.3
prompt-toolkit==3.0.7
protobuf==3.13.0
psutil==5.7.2
ptyprocess==0.6.0
pyarrow==1.0.1
pycparser==2.20
pydeck==0.5.0b1
Pygments==2.6.1
pyparsing==2.4.7
pyrsistent==0.17.3
python-dateutil==2.8.1
pytz==2020.1
PyYAML==5.3.1
pyzmq==19.0.2
regex==2020.7.14
requests==2.24.0
s3transfer==0.3.3
sacremoses==0.0.43
scikit-learn==0.23.2
scipy==1.5.2
Send2Trash==1.5.0
sentencepiece==0.1.91
sentry-sdk==0.18.0
seqeval==0.0.18
shortuuid==1.0.1
simpletransformers==0.51.4
six==1.15.0
smmap==3.0.4
streamlit==0.67.1
subprocess32==3.5.4
tensorboard==1.15.0
tensorboardX==2.1
tensorflow==1.15.0
tensorflow-estimator==1.15.1
termcolor==1.1.0
terminado==0.9.1
testpath==0.4.4
threadpoolctl==2.1.0
tokenizers==0.9.4
toml==0.10.1
toolz==0.11.1
torch==1.6.0+cu101
torchvision==0.7.0+cu101
tornado==6.0.4
tqdm==4.48.2
traitlets==5.0.3
transformers==4.1.1
tzlocal==2.1
ujson==4.0.1
urllib3==1.25.10
validators==0.18.1
wandb==0.10.4
watchdog==0.10.3
wcwidth==0.2.5
webencodings==0.5.1
Werkzeug==1.0.1
widgetsnbextension==3.5.1
wrapt==1.12.1
zipp==3.3.0

**Comment by ThilinaRajapakse** [comment]
Off the top of my head, setting the seed with the same hyperparameters should give the same result. How big of a difference are you seeing?
